### ISSUE TITLE
Deal with comments

### ISSUE THREAD

**Issue by mdboom** [spacetelescope/asdf-standard#50]
Currently, most YAML parsers just throw comments away, so they can't be round tripped.

Preserving comments is tricky -- you have to both modify the YAML parser and modify the objects returned so they have an extra place to stick the comments, which creates an asymmetry between the objects passed into ASDF for saving (which are likely "native" objects), and object you get out from parsing (which would be some subclass of the native objects with an extra comment hanging off the side).

An alternative is to create some convention for attaching comments to an arbitrary mapping, something like how arbitrary references can be used using the JSON Pointer standard.

Anyway, here's a thread for discussing options...

Cc: @embray @perrygreenfield

**Comment by embray** [comment]
What I did in pyfits to avoid the "subclass of native objects" difficulty was just to add a `header.comments` mapping that lets you look up the comment associated with a given keyword.  Likewise there could be a `tree.comments` that can look up comments by JSON Pointer reference.  A bit of a pain, maybe, but an option.  Either the schema, or the validator itself, would have to be updated to always accept these special commentary strings.

**Comment by mdboom** [comment]
That approach works for FITS, where the key/value pairs are flat, but in YAML things might get tricky.

Let's say you have a `tree` to store the regular data and other data structure `comments` that is say a mapping from JSON Pointer paths to comments.   If the tree is restructured in any way, the comments may no longer point to the right thing.  That's why I had suggested the need for subclasses of basic types...

An alternative suggestion is to "suggest" a convention for storing comments on any mapping.  To reduce verbosity, there's actually a number of single-character keys we could use:

```
foo:
  $: my comment
  //: my comment
  ^: my comment
  -: my comment
  =: my comment
  (: my comment
  <: my comment
  +: my comment
  _: my comment
```

I'm kind of partial to `//` as it's the comment token from C-like languages.  `#` and `%` won't work, for obvious reasons.

**Comment by perrygreenfield** [comment]
Should we just remove the whole aspect of comments from the paper? I wonder if it is the wrong concept to propagate anyway. I would almost favor saying that if you want some statement permanently associated with an item in the tree, then it would get a standard keyword like "description" (implying only objects can have descriptions?).  And, yes, even raising the issue of trying to use YAML comments in any way is a waste of time if it isn't part of the standard. Actually, it may be worth it to say we discourage their use. Another approach may be to imply any descriptions on standard objects reside in schema's and provide tools to display such descriptions as part of special Tree displays? After all, why clutter up Trees with such stuff if it is standard? Shouldn't the schema be the right place for such stuff?

**Comment by mdboom** [comment]
I tend to agree with Perry overall.

Many FITS comments describe the purpose of the key -- and for ASDF, it's fine to put that information in one place, in the schema.  But it's *possible* (though maybe rarely used in practice, I guess???) for the comment to be referring to the value, perhaps to describe a numerical instrument mode, for example.

To be clear (since I wasn't), I was thinking of comments more in the sense of notes that one leaves for oneself and others to explain why something is a certain way (a la comments in source code).  So you might have a comment "I chose mode X here rather than the more obvious mode Y because mode Y is limited in the Z dimension".  If you used a regular YAML comment for this, it wouldn't roundtrip and would disappear if the file were updated by some manual process.  If you used a convention like what I propose above, it could be -- and the ASDF standard would say clearly that tools should retain this key, but should never perform any sort of action based on it: it is a comment, after all.

As for the paper, I don't know if we need to get into the weeds about all this, though.

**Comment by embray** [comment]
> But it's possible (though maybe rarely used in practice, I guess???) for the comment to be referring to the value, perhaps to describe a numerical instrument mode, for example

Although even this information can be captured in the schema, so I think that's the right place for most such information.

I like the idea of suggesting a convention though.   I also like `//` out of all the choices (I also like `>` but it won't work without quotes around it :/)

**Comment by mdboom** [comment]
Closed by #59.
